### PR TITLE
[RFC] Add fuzzer for Python code using llvm-libfuzzer

### PR DESCRIPTION
**Pull request by MarkusTeufelberger**
This is not yet styled according to the guidelines or feature complete, more of a proof of concept that I'd like some feedback on.

As https://github.com/micropython/micropython/pull/1884 ran into the issue that byte code is assumed to be correct, this will first compile any input (making the fuzzer slower of course) and then execute the result. The benefit is that the compiler is fuzzed too, the downside is the slowdown due to compilation. It is however possible to write a second fuzzer with the same instrumentation to explicitly fuzz just bytecode.

ASAN and UBSAN do really not like your code as your memory management is more "hands-on" and I have to research more into what checks to disable to still make them trigger on useful errors. For now, only LSAN is enabled.
I also likely did stuff horribly wrong in python_fuzzer.cc (e.g. not setting up sys.argc/sys.argv properly) and it would be great if someone could look over it and tell me where I'm wrong or how to improve.

The basic idea of libfuzzer is to provide some random bytes that then get used as input by whatever is being fuzzed. Similar to AFL, libfuzzer then will check which parts of the code were active in processing the input and then modify it accordingly, creating the next input.

It is possible to bootstrap the fuzzer using the existing test cases, at the moment I seem to fail a couple of them on my machine though, so the more radical approach of just letting the fuzzer figure out the whole syntax on its own is used.

Unlike AFL, libfuzzer will stop once a timeout or crash is reached. The -jobs parameter decides how many such fragments will be collected until the process ends. Timeouts will happen frequently while fuzzing randomly mutated input into a programming language, since already the small program "9**99999" will run for quite a while.

"make consolidate" will run over the existing corpus and reduce it to useful test cases, so if you later provide test cases with very high coverage manually, you can still keep using fuzzed results but will get rid of a lot of tiny noisy test cases.

"make coverage" should in the future generate a html report similar to lcov. Maybe it might be a better approach to just build an actual gcov enabled binary and do "proper" coverage reporting with lcov, I didn't find a lot of eye candy options for sancov, also the html stuff is a relatively new feature.

Looking forward to your comments! :)

### CODE REVIEW

**Comment by dpgeorge** [comment]
Thanks for the contribution!  But it might be some time before this can get a proper look, due to other commitments.

**Comment by MarkusTeufelberger** [comment]
Alright, now it should be able to use (most) of the .py files in /tests as seed files to get a head start in fuzzing.

Just to make sure anyone trying this out or reviewing it doesn't get hurt: This is a fuzzer that will run as much randomly generated code as possible - which can include ANYTHING that is valid Python on your machine. Including file system operations for example... Please run this in a VM, container or similar sandboxed environment.

Also there was a bit of confusion about me "dumping" a large feature onto you: As it says in the title, I'm not proposing to merge this, I'm rather using this PR to ask if this is the correct way to initialize a minimal interpreter (in python_fuzzer.cc) instead of opening a thread/ticket for every little problem I run into. Currently it crashes on some input that the unix micropython executable handles just fine, though I've also found a few cases that crash both reliably. If it finally is working correctly, I would offer to properly write a patch that either integrates in your main Makefile or resides in the tests folder, with proper commit messages, documentation etc.

**Comment by pfalcon** [comment]
> Currently it crashes on some input that the unix micropython executable handles just fine, though I've also found a few cases that crash both reliably.

Can you shows those inputs?

> this PR to ask if this is the correct way to initialize a minimal interpreter (in python_fuzzer.cc)

Before someone finds time to look into a detail of it, did you try to look at
https://github.com/micropython/micropython/blob/master/minimal/main.c ? That's definitely a correct way to initialize a *minimal* interpreter. And if you want to fuzz something, you probably want to start with that thing, which can't write a filesystem, etc.

**Comment by MarkusTeufelberger** [comment]
> Can you shows those inputs?

E.g. `7*99* '*297*"'*9*5*3714730` tries to allocate ~3.5 GB of memory and fails with a MemoryError for me. I went through a few others, but they are not very interesting and so far I didn't really focus on keeping all resulting artefacts. This one for example looks suspiciously like it is shooting itself in the foot, ending with a segfault (minimized manually):

    import uctypes
    desc = {"arr3": (uctypes.ARRAY | 2, uctypes.UINT16 | 2)}
    data = bytearray(0)
    S = uctypes.struct(uctypes.addressof(data), desc, uctypes.LITTLE_ENDIAN)
    S.arr3[0] = 0x2233

Once it works properly, I'll run it a few hours to find a few more interesting examples

> did you try to look at [...]

Yes, see line 29 in python_fuzzer.cc.
Since there was a big warning in the comments about the GC potentially not working correctly and the sys.argv test was constantly failing, I also used a bit of code from the unix main.c file.

What I essentially need is the do_str() function + the necessary boilerplate code around it so it actually works.

**Comment by pfalcon** [comment]
> tries to allocate ~3.5 GB

Well, that's definitely a bug, it should allocate only within a uPy heap (set to 1MB).

> This one for example looks suspiciously like it is shooting itself in the foot, ending with a segfault (minimized manually):

uctypes is known to lack boundary checks, and it's kinda by design - it's so powerful anyway, that following GIGO principles for as long as there're better things to do makes sense.
https://github.com/micropython/micropython/issues/1761

**Comment by pfalcon** [comment]
Another idea: instead of trying to fuzz uPy (I still don't understand what drives you and what you want to achieve, following my metaphors, you're trying to hit a car with a brick - yes, there will be damage), try start with fuzzing libs uPy uses: re1.5, uzlib, berkeley-db-1.85 I've just added.

**Comment by MarkusTeufelberger** [comment]
I'm really trying to not discuss or deliver bugs found with this here to not derail the discussion. Following your metaphor: When your car is hit by a brick in the door, does it have a dent or does it explode? Do you manually poke it with a stick every once in a while to see if it is still stable or do you see value in putting a banana inside and letting a few monkeys try to get in to see if the doors are not suddenly replaced by a sheet of paper?

These libs seem to be focussing on size, not robustness. Also you seem to be not really convinced of this whole idea, so I'm kinda hesitant to do work for your private side projects or database engines from the 90s.

**Comment by pfalcon** [comment]
> These libs seem to be focussing on size, not robustness. Also you seem to be not really convinced of this whole idea, so I'm kinda hesitant to do work for your private side projects or database engines from the 90s.

These libs are part of micropython and were factored out for reuse, and people slowly leverage them, so working on them supposedly would benefit more people. Yes, they are focused on size, but by now I'm pretty sure they approaching that stage in their lifetime when their users don't want to get a crash because "wrong" gzip stream was fed in.

My being "not really convinced" is based on the fact that I'm working on uPy in my free time, and in no way I'm spending my free time on twists of Random, if instead I can spend it using powers of Reason to move straight at the needed direction (which so far functionality, not brick-proofness). On the other hand, if I were hired to work on uPy, and an employer wanted to make uPy more robust, your work would be spot on, and we'd cooperate. So, as I hinted above, it would be helpful to know what drives you to do that.

And regarding Berkeley db - it may be a "database engine from 90ies", but I hope you know that algorithms are timeless, and bdb 1.xx is almost raw implementation of an algorithm, devoid of usual layers and layers of crap of "modern software". That's exactly what uPy needs. And it indeed was a big thing in 90ies. But building it with standard uPy warnings, in a dozen of usual false positives, I actually found one bug and 2 portability issues (bugs too). So, assuming you're a practicing security researcher sharpening your skills, vivisecting bdb would be a lot of fun, in 90ies big corporations went for it, but now you with a day of fuzzing found a lot of bugs in it, showing what a stone age these 90ies were, with people living in caves. Sounds fun.

It's of course up to you what to work on and why.

**Comment by pfalcon** [comment]
> 7*99* '*297*"'*9*5*3714730 tries to allocate ~3.5 GB of memory and fails with a MemoryError for me.

Well, your description made it sound like it thrashed your swap for half and hour, just to report MemoryError at the end. Nope, it just computes amount of memory required to hold the result, and fails right away. It's not even a DoS case, as nothing gets really allocated. So, it works perfectly. The only issue you could raise is that there looks to be a 32-bit wrap-around somewhere, so it's possible to trick "a" * <bignumber> to produce just "a". Feel free to prove that and submit it, and at least from me it'll get "almost non-bug" status, so I'll let it ripe for at least a year before looking into it - we have more important issues waiting longer. (And that proves my point - fuzzing may be fun (hmm), but dealing with its results (99.9999% false positives or "almost false positives") is deadly boring).

**Comment by MarkusTeufelberger** [comment]
> I'm really trying to not discuss or deliver bugs found with this here to not derail the discussion.

**Comment by pfalcon** [comment]
Any news on this? I see that https://github.com/renatahodovan/fuzzinator finds in another project a whole bunch of bugs, whoa! So either that project is particularly buggy or I should eat my hat and fuzzing is not that useless (or that particular tool is way cool). Indeed, bugs like https://github.com/micropython/micropython/issues/2886 could realistically be caught with fuzzing.

**Comment by MarkusTeufelberger** [comment]
If you are interested in fuzz testing, I can recommend asking to be included at https://github.com/google/oss-fuzz - for creating the actual fuzzer I am not sure if libfuzzer is the perfect tool for that, at least not for fuzzing the whole language. Already evaluating very short statements like 9**99999 takes quite long, but that is only 8 characters of input. Maybe Google is fine with dealing with these timeout issues or workers being busy doing essentially nothing while using a full CPU, maybe not.

They are at least interested in fuzzing (c)Python (https://github.com/google/oss-fuzz/issues/402), so maybe you can look for inspiration/approaches there too.

Fuzzinator seems to be a framework to do fuzz testing and getting the reports etc., not so much the fuzzer itself. libFuzzer is a instrumentation guided fuzzer, meaning it will try to change its input in a way to hit as many code paths as possible, checking after every test file if a different/unknown code path was used and acting accordingly. The artifacts generated then would be probably categorized, reported upon etc. by fuzzinator (it seems to include an AFL runner for example, which works very similar to libFuzzer). Fuzzinator would be more comparable to ClusterFuzz, Google's closed source internal solution for managing fuzzers. OSS-Fuzz is the common framework to provide them with the inputs for building and running these at scale.

As my initial question for feedback still is waiting for someone to actually look at the few dozen lines of copy-pasted code and telling me if I am invoking lexer, parser, compiler + setup/teardown boilerplate correctly to essentially run eval() on a given char*, I haven't done further work on this since.

There is of course some work to be done to make this faster and get it actually in a useful state (e.g. moving some parts out of the inner loop), but having a correct eval() is the first step.

Another complication as far as I remember was (is?) that µpython seems to be designed to do lots of its memory management itself in a simplified and sometimes more streamlined way, something that e.g. AddressSanitizer or UndefinedBehaviourSanitizer do not really like. This means it might be a bit harder to get the program to crash with a detailed report as soon as possible.

File: tests/libfuzzer/python_fuzzer.cc [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    // Heap size of GC heap (if enabled)
+    // Make it larger on a 64 bit machine, because pointers are larger.
+    long heap_size = 1024*1024 * (sizeof(mp_uint_t) / 4);
+    char *heap = (char*)malloc(heap_size);
+    gc_init(heap, heap + heap_size);
+    #endif
+
+    mp_init();
+
+    char *fakeargv[] = {"fuzzer", "foo", "bar"};
```
Comment (dpgeorge) [inline tests/libfuzzer/python_fuzzer.cc]: You'll need to add: `mp_obj_list_init(mp_sys_path, 0);`

File: tests/libfuzzer/python_fuzzer.cc [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    mp_obj_list_init((mp_obj_list_t*)MP_OBJ_TO_PTR(mp_sys_argv), 0);
+    for (int i = 0; i < 3; i++) {
+        mp_obj_list_append(mp_sys_argv, MP_OBJ_NEW_QSTR(qstr_from_str(fakeargv[i])));
+    }
+
+    mp_lexer_t *lex = mp_lexer_new_from_str_len(MP_QSTR__lt_stdin_gt_, (const char*)data, size, 0);
```
Comment (dpgeorge) [inline tests/libfuzzer/python_fuzzer.cc]: Check if lex is NULL and exit with an error if it is.

File: tests/libfuzzer/python_fuzzer.cc [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+    nlr_buf_t nlr;
+    if (nlr_push(&nlr) == 0) {
+        qstr source_name = lex->source_name;
+        mp_parse_tree_t parse_tree = mp_parse(lex, MP_PARSE_FILE_INPUT);
+        mp_obj_t module_fun = mp_compile(&parse_tree, source_name, MP_EMIT_OPT_NONE, true);
```
Comment (dpgeorge) [inline tests/libfuzzer/python_fuzzer.cc]: The "true" arg should be "false" (the input does not come from a REPL, especially if it's of the MP_PARSE_FILE_INPUT form).  Actually, even better would be to fuzz this argument, along with the second arg to mp_parse.

File: tests/libfuzzer/python_fuzzer.cc [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    //Initialize stuff (?)
+    mp_stack_ctrl_init();
+
+    mp_stack_set_limit(40000 * (BYTES_PER_WORD / 4));
+
+    #if MICROPY_ENABLE_GC
+    // Heap size of GC heap (if enabled)
+    // Make it larger on a 64 bit machine, because pointers are larger.
+    long heap_size = 1024*1024 * (sizeof(mp_uint_t) / 4);
+    char *heap = (char*)malloc(heap_size);
```
Comment (dpgeorge) [inline tests/libfuzzer/python_fuzzer.cc]: Unless you want to change the heap size on the fly, it's simpler to just create a static buffer (stored in the BSS, eg static uint8_t heap[1024*1024]) and use that for the heap.

**Comment by dpgeorge** [comment]
Closing due to inactivity.  See also #3798.

**Comment by vincedani** [comment]
Hi @MarkusTeufelberger, it looks like you were fuzzing this repository based on this pull request. Although, it’s a quite old and closed PR, I would like to ask whether you have had found some bugs? 

I would like to create a Python based test case reduction test suite that contains fuzzer generated outputs, and benchmark automatic test case reducers how they perform on Python inputs. Is it possible that you have some bugs with the output of the fuzzer, which is free of any reduction?

**Comment by MarkusTeufelberger** [comment]
Hi @vincedani,

No, this was one of my first tries to add fuzzing to a code base and in hindsight probably both too naive and too ambitious.

I didn't get much deeper than the code here and the approach has the obvious problems stated above (very short inputs can take very long to evaluate and yet be completely valid anyways, so checking for bugs in your timeouts is not helpful - and once it discovers the `**` operator, you can expect to run into a LOT of timeouts). I'm not sure if general programming languages are even well fuzzable with an approach like this since the computation and output space is so large. You might need to stop after the compiler step (as `eval()` is too potentially complex), but then you're also not very deep into what you actually probably want to achieve and are just fuzzing kinda the initialization infrastructure around the language instead of the actual computations or libraries.

By the way, I wrote a test case minimizer myself too, in case you want to take a look or add it to your benchmarking suite: https://github.com/MarkusTeufelberger/afl-ddmin-mod

**Comment by vincedani** [comment]
Thanks @MarkusTeufelberger for the information, I'll take a look the referenced repository.

**Pull request closed without merging**